storage/innobase/fts/fts0opt.cc

```diff
@@ -1423,87 +1401,63 @@ fts_optimize_word(
-	dberr_t		error = DB_SUCCESS;
-	char		table_name[MAX_FULL_NAME_LEN];
-
-	info = pars_info_create();
-
-	ut_ad(fts_table->charset);
-
-	pars_info_bind_varchar_literal(
-		info, "word", word->f_str, word->f_len);
-
-	selected = fts_select_index(fts_table->charset,
-				    word->f_str, word->f_len);
-
-	fts_table->suffix = fts_get_suffix(selected);
-	fts_get_table_name(fts_table, table_name);
-	pars_info_bind_id(info, "table_name", table_name);
-
-	graph = fts_parse_sql(
-		fts_table,
-		info,
-		"BEGIN DELETE FROM $table_name WHERE word = :word;");
-
-	error = fts_eval_sql(trx, graph);
-
-	if (UNIV_UNLIKELY(error != DB_SUCCESS)) {
-		ib::error() << "(" << error << ") during optimize,"
-			" when deleting a word from the FTS index.";
-	}
-
-	que_graph_free(graph);
-	graph = NULL;
-
-	/* Even if the operation needs to be rolled back and redone,
-	we iterate over the nodes in order to free the ilist. */
-	for (i = 0; i < ib_vector_size(nodes); ++i) {
-
-		fts_node_t* node = (fts_node_t*) ib_vector_get(nodes, i);
-
-		if (error == DB_SUCCESS) {
-			/* Skip empty node. */
-			if (node->ilist == NULL) {
-				ut_ad(node->ilist_size == 0);
-				continue;
-			}
-
-			error = fts_write_node(
-				trx, &graph, fts_table, word, node);
-
-			if (UNIV_UNLIKELY(error != DB_SUCCESS)) {
-				ib::error() << "(" << error << ")"
-					" during optimize, while adding a"
-					" word to the FTS index.";
-			}
-		}
-
-		ut_free(node->ilist);
-		node->ilist = NULL;
-		node->ilist_size = node->ilist_size_alloc = 0;
-	}
+  FTSQueryExecutor executor(trx);
+  ulint selected = fts_select_index(fts_table->charset, word->f_str,
+                                    word->f_len);
+  fts_table->suffix = fts_get_suffix(selected);
+  dict_table_t *aux_table= nullptr;
+
+  dberr_t err= executor.open_table(fts_table, &aux_table);
+  if (err) goto free_node;
+
+  executor.prepare_for_write(aux_table);
+  executor.build_tuple(aux_table, 1, 1);
+  executor.assign_aux_table_fields(word, nullptr);
+  err= executor.delete_rec(aux_table);
+
+  if (err == DB_RECORD_NOT_FOUND || err == DB_END_OF_INDEX)
+    err= DB_SUCCESS;
+
+  if (UNIV_UNLIKELY(err != DB_SUCCESS))
+    ib::error() << "(" << err << ") during optimize,"
+                   " when deleting a word from the FTS index.";
+free_node:
+  /* Even if the operation needs to be rolled back and redone,
+  we iterate over the nodes in order to free the ilist. */
+  for (uint32_t i = 0; i < ib_vector_size(nodes); ++i)
+  {
+    fts_node_t* node = (fts_node_t*) ib_vector_get(nodes, i);
+    /* Skip empty node. */
+    if (node->ilist == NULL)
+    {
+      ut_ad(node->ilist_size == 0);
+      continue;
+    }
 
-	if (graph != NULL) {
-		que_graph_free(graph);
-	}
+    if (err == DB_SUCCESS)
+    {
+      err = fts_write_node(aux_table, word, node, &executor);
 
-	return(error);
+      if (UNIV_UNLIKELY(err != DB_SUCCESS))
+        ib::error() << "(" << err << ") during optimize, while adding a"
+		    << " word to the FTS index.";
```

Comment:
Can we please use `sql_print_error` in new code?

The last `operator<<` invocation is redundant here; it would concatenating two compile-time constant strings at runtime.